Homological Discrete Vector Fields

umenohana13 opened the pull request:
## Summary of Changes

The pull request contains the first version of:
* user manual (still incomplete - persistant homology and Alexander duality are not documented)
* reference manual (slightly incomplete as well - the IO classes stores in tools_io are "lightly" documented as they will be revised to integrate IOs inside CGAL)
* code (same: the tools_io file will be revised)

## Release Management

* Affected package(s): `HDVF` / `Surface_mesh_topology`
* Issue(s) solved (if any): 
* Feature/Small Feature (if any): https://cgalwiki.geometryfactory.com/CGAL/Members/wiki/Features/Homological_Discrete_Vector_Fields
* License and copyright ownership: Not yet validated by my lab but most probably LGPL / LIS laboratory

afabri commented:
/force-build:v0

afabri commented:
/force-build:v0

afabri commented:
/build:v1

umenohana13 commented on HDVF/doc/HDVF/Concepts/AbstractChainComplex.h:
```diff
@@ -0,0 +1,132 @@
+/*!
+Returns the dimension of the complex, that is, the largest dimension of cells.
+ */
+int dim();
```
Done

umenohana13 commented on HDVF/doc/HDVF/Concepts/Filtration.h:
```diff
@@ -0,0 +1,165 @@
+
+The concept `Filtration` describes the requirements for persistent filtrations associated to persistent homology computation.
+
+A filtration is associated to an `AbstractSimplicialComplex`. Each cells is equiped with a scalar value (called its degree) and the filtration is an enumeration of cells in any dimension by increasing degrees.
```
Oups, it's "AbstractChainComplex". 
DONE

umenohana13 commented on HDVF/doc/HDVF/Concepts/Filtration.h:
```diff
@@ -0,0 +1,165 @@
+
+ - an iterator to visit all cells by increasing degrees.
+ - getters to get the degree of a cell, and the cell of a given index along the filtration.
+ - an overload of `<<` to output filtrations.
```
DONE

afabri commented:
/force-build:v1

afabri commented on HDVF/include/CGAL/HDVF/Hdvf_core.h:
```diff
@@ -0,0 +1,1172 @@
+    {
+        // Update _G_col[q + 1]
+        // Subtract the product of (G11 * D11_inv) and D12 from _G_col[q+1]
+        _G_col[q + 1] -= (G11 * D11_inv) * D12;
```
@sloriot now that you master VC++, can you please have a look.

umenohana13 committed:
Fixes

-> removed CoefficientType from tparams in all classes (except in AbstractChainComplex models)
-> added a Coefficient_type definition in AbstractChainComplex (to get the taparm)
-> added a typedef for ComplexType::Coefficient_type in all classes
-> replaced references to Ring by EuclideanRing (which already belong to CGAL ... it's more technical than what we needed but I think it's OK)
-> mentionned that DegreeType must be RealEmbeddable

umenohana13 committed:
Merge branch 'Homology-new_package-AB' of github.com:umenohana13/mine into Homology-new_package-AB

afabri commented:
/force-build:v1

afabri committed:
Update HDVF/include/CGAL/HDVF/Geometric_chain_complex_tools.h

afabri commented on HDVF/include/CGAL/HDVF/Geometric_chain_complex_tools.h:
```diff
@@ -0,0 +1,474 @@
+
+The class `Duality_cubical_complex_tools` is dedicated to Alexander duality for 3D binary volumes.
+
+Starting from a Cubical_chain_complex `_K`, the method `cubical_chain_complex_BB` builds a Cubical_chain_complex L and Sub_chain_complex_mask K.
```
```suggestion
Starting from a Cubical_chain_complex `_K`, the method `cubical_chain_complex_bb()` builds a Cubical_chain_complex L and Sub_chain_complex_mask K.
```

afabri replied:
I think it would read nicer if you replaced Cubical_chain_complex with cubical chain complex. In case you want to keep it you must backtick, as it is a type.  Why the letter "K"?  Can you avoid the starting underscore?

afabri replied:
Also the different classes should better be in their own header files.

umenohana13 commented on HDVF/include/CGAL/HDVF/Hdvf_persistence.h:
```diff
@@ -0,0 +1,839 @@
+    // Copy _DD_col with filtration order (for dimensions q>0)
+    for (int q = 0 ; q <= this->_K.dimension(); ++q)
+    {
+        const std::pair<int, int> s(this->_DD_col.at(q).dimensions()) ;
```
Thanks, corrected

umenohana13 commented on HDVF/include/CGAL/HDVF/Hdvf_persistence.h:
```diff
@@ -0,0 +1,839 @@
+}
+
+template<typename CoefficientType, typename ComplexType, typename DegreeType, typename FiltrationType>
+Cell_pair Hdvf_persistence<CoefficientType, ComplexType, DegreeType, FiltrationType>::step_persist(bool& found, bool verbose)
```
OK, corrected

umenohana13 committed:
Fixes

-> corrected step_persist to return a pair
-> started correcting tparams and type names

umenohana13 committed:
Merge branch 'Homology-new_package-AB' of github.com:umenohana13/mine into Homology-new_package-AB

umenohana13 committed:
Fixes

Fixed tparam and typedef names

umenohana13 committed:
Rework of Hdvf_persistence types

-> removed all unnecessary types
-> created a struct PerHole
-> ajusted the code accordingly
-> changed the encoding of infinite holes (set birth = death, simpler...)

umenohana13 committed:
Hdvf_persistence rework

-> Finished reworking types in Hdvf_persistence (removal of all the unnecessary "small" types -> creation of a unique PerHole structure)

TODO: adjust the times to avoid camelcase...

umenohana13 commented on HDVF/include/CGAL/HDVF/Cubical_chain_complex.h:
```diff
@@ -0,0 +1,1081 @@
+     */
+    Cubical_chain_complex(const Cub_object_io& cub,typeComplexCube type);
+
+    /** \brief Friend class `Duality_cubical_complex_tools` provides tools for Alexander duality. */
```
Yes I needed these classes to be friend as the tools for duality really manipulate complexes (and private parts).
I'll check but I'm pretty sure I needed this friendship

But the one I needed more was in Sparse_chain and Sparse_matrix to overload operators "symmetrically" ... I hope it's OK for you...

umenohana13 replied:
I confirm that Duality classes are very technical and it was more convenient to make them friend (as they produce new chain complexes + chain complex masks starting from the parent chain complex)

afabri replied:
I had not checked whether friendship is needed or not.  I commented as I rarely saw so many friend declarations in a package, and this might be an indication to better add public functions.

umenohana13 replied:
OK I'll check again to see if I can improve public functions and remove friends.
A question: for Sparse_chain and Sparse_matrix, there are a lot of friends because I was told that operators are better overloaded using friends (in order to define "symmetrical" operators). Are friends OK in this setting?

umenohana13 replied:
OK - I trust you. So if you think operators should not be friends I will modify Sparse_chain and Sparse_matrix accordingly. 
I add it to my TODO list

umenohana13 replied:
I removed all friends in Cubical_chain_complex and Simplicial_chain_complex. You were right: I lacked some public methods. Now there's no friendship anymore between chain complex methods and Geometric_chain_complex_tools

umenohana13 replied:
I limited friends as much as possible. All the remaining friends use private or protected members and are "friends" because they provide optimised overloads depending on chain/matrix types (so that the choice of the right method is guided by the ChainTypeFlag).

umenohana13 committed:
For Andreas (problem with << overload)

umenohana13 commented on HDVF/include/CGAL/HDVF/Cubical_chain_complex.h:
```diff
@@ -0,0 +1,1081 @@
+     *
+     * \param[in] complex The cubical chain complex which will be copied.
+     */
+    Cubical_chain_complex& operator=(const Cubical_chain_complex& complex)
+    {
+        _dim = complex._dim;
```
I think it's OK (data member initialised by a method called on an argument cannot be initialised in the : ?

umenohana13 commented on HDVF/include/CGAL/HDVF/Cubical_chain_complex.h:
```diff
@@ -0,0 +1,1081 @@
+    static const std::vector<int> VTK_cubtypes ;
+
+public:
+    /** \brief Type used to encode primal or dual construction. */
```
This enum is also very technical (it contains the codes used to encode respectively vertices, edges, squares and cubes in VTK). So normally users should not need it

umenohana13 committed:
Fixes

Fixes related to persistent homology VTK export

umenohana13 commented on HDVF/include/CGAL/HDVF/Geometric_chain_complex_tools.h:
```diff
@@ -0,0 +1,474 @@
+ */
+
+template <typename ComplexType, template <typename, int> typename _ChainType = OSM::Sparse_chain, template <typename, int> typename _SparseMatrixType = OSM::Sparse_matrix, typename VertexIdType = size_t>
+void hdvf_geometric_chain_complex_output_vtk (Hdvf_core<ComplexType, _ChainType, _SparseMatrixType> &hdvf, ComplexType &complex, std::string filename = "test", bool co_faces = false)
```
OK so I should move these functions to the IO namespace? 
I'm currently trying to revise all these classes. So I'll modify as you suggest

afabri committed:
Update HDVF/include/CGAL/HDVF/Hdvf.h

umenohana13 committed:
Fixes

-> renamed HDVF VTK output methods to write_VTK and moved them to CGAL::IO
-> Corrected tparams / names of types in Hdvf_persistence and Hdvf_duality
-> Fixed a regression in Tet_object_io and Hdvf_duality
-> Added example_dual_hdvf_simplicial

albert-github commented:
@afabri Looks like the documentation didn't succeed as https://cgal.geometryfactory.comcould not be reached, looks like the machine / connection is down.

umenohana13 committed:
Fixes

-> Removed unnecessary nodes argument in Simplicial_chain_complex (Mesh_object_io) constructor

umenohana13 committed:
Cubical_chain_complex fixes

-> after removal of friends in Cubical_chain_complex, simplification of public/protected methods

umenohana13 committed:
Fix

Removed commented code

afabri commented:
/force-build:v1

afabri commented on HDVF/doc/HDVF/Concepts/Ring.h:
```diff
@@ -0,0 +1,50 @@
+\ingroup PkgHDVFConcepts
+\cgalConcept
+
+The concept `Ring` describes the requirements for the ring of coefficients used to compute homology in the `HDVF` concept. Besides ring operators, it also specifies the functions needed to test invertibility in the ring.
```
I am wondering if the `Ring` concept is identical to [`IntegralDomainWithoutDivision`](https://doc.cgal.org/latest/Algebraic_foundations/classIntegralDomainWithoutDivision.html).
If it is different this concept gives an idea what you have to add to the new concept.

afabri commented on HDVF/doc/HDVF/Concepts/SparseChain.h:
```diff
@@ -0,0 +1,305 @@
+    SparseChain();
+
+    /*!
+     * \brief Creates new empty sparse chain of given size.
```
It is not empty.  Do you mean that all coefficients have a particular value?

afabri committed:
Update HDVF/include/CGAL/HDVF/Hdvf_core.h

typo

afabri commented on HDVF/include/CGAL/HDVF/Hdvf_core.h:
```diff
@@ -0,0 +1,1213 @@
+ Homology/cohomology generators are actually algebraic objects, namely chains. Methods `homology_chain()` and `cohomology_chain()` return the homology and cohomology generator chain associated to a given critical cell. VTK export functions output all the cells of such chains with non zero coefficients.
+
+
+ \cgalModels{HDVF}
```
This links to namespace and not to the concept. Maybe @sloriot can help.

umenohana13 committed:
Friends rework

-> removed unnecessary friends in Sparse_chain

umenohana13 committed:
Friends rework

-> removed unnecessary friends in Sparse_chain and Sparse_matrix
-> added \relates for remaining ones
-> fixed write_VTK for persistent HDVFs (call of write_VTK HDVF_core)

umenohana13 committed:
Friends rework + fixes

-> added all \relates
-> removed unnecessary friends
-> fixed [] (now private) calls in Sparse_matrix

umenohana13 committed:
Inline fixes

- inlined set_coefficient / get_coefficient in Sparse_chain and Sparse_matrix (for efficiency)

afabri commented:
/force-build:v1

umenohana13 commented on HDVF/include/CGAL/HDVF/Hdvf_core.h:
```diff
@@ -0,0 +1,1213 @@
+namespace CGAL {
+namespace HDVF {
+
+/** \brief HDVF Enum for the label of cells. */
```
I fully agree. 
Before the zoom, some tries  (usually in the papers we say that P, S or C is a label or that we define a partition of cells into primary, secondary, critical):
- HDVFFlag
- HDVFLabel
- HDVFPartitionLabel
I think YOU should tell me (I'm too much used to HDVFs to be clear on this)
-

umenohana13 commented on HDVF/include/CGAL/HDVF/Hdvf_core.h:
```diff
@@ -0,0 +1,1213 @@
+
+ The class `Hdvf_core` is the core implementation of homological discrete vector fields (HDVF for short). The ring of coefficients for homology computation must be a model of `Ring`.
+
+ An enumeration `Flag_type` is defined in the `HDVF` namespace and the `Hdvf_core` class maps each cell to one of the flags (namely `PRIMARY`, `SECONDARY`, `CRITICAL`). The NONE flag is used in child classes (such as `Hdvf_duality` or `Hdvf_reduced`) when computing relative homology on a sub-complex.
```
You're right ... actually "not yet" exist (I was short of time). But it's a straightforward adaptation of Hdvf_duality that can be interesting for users (reduced homology is pretty much used in practice while no software provide its computation...). 
For now I removed it ... later :)

umenohana13 commented on HDVF/include/CGAL/HDVF/Hdvf_persistence.h:
```diff
@@ -0,0 +1,821 @@
+     *
+     * \returns A vector containing, for each dimension, the vector of labels by cell index.
+     */
+    virtual std::vector<std::vector<int> > psc_labels () const
```
I thought I did not use polymorphism, BUT it seems I use some (as I noticed that if I remove the virtual of "flag" method in Hdvf_core, I meet problems for Hdvf_duality - where only flags of the current sub_chain_complex_mask are exported).
I suggest that I write tests for HDVFs quickly and that I use these tests then to remove all unnecessary virtual (anyway all virtual in Hdvf_duality and Hdvf_persistance are useless - and just come from rough "copy" ...)

umenohana13 committed:
Fixes

Fixes following "features" discussions and github comments

umenohana13 committed:
Surface_mesh_io

New (prototype) class to load a CGAL::Surface_mesh:
-> Surface_mesh_io (inherits Mesh_object_io and adds internal structures to keep a connection between simplex (Io_cell_type at this stage) <-> Halfedge_index
-> provide a basic test_surface_mesh_io

afabri committed:
Update HDVF/include/CGAL/HDVF/Hdvf_core.h

umenohana13 committed:
Fixes

-> Renamed HDVF flag (previously Flag_type) into PSC_flag
- getters are now called psc_flags (getting several flags) and psc_flag (for a single cell)
-> Renamed persistent data coherently:
- Persistent_interval now related to "time and degree" informations (so the "interval" data
- Persistent_hole includes a Persistent_interval and homology/cohomology generators + PSC_flags
-> Renamed chain type flag -> StorageFormat in Sparse chains and matrices

umenohana13 committed:
Merge branch 'Homology-new_package-AB' of github.com:umenohana13/mine into Homology-new_package-AB

umenohana13 committed:
Fix

Changed namespace HDVF to Homological_discrete_vector_field

afabri commented on HDVF/include/CGAL/HDVF/Simplex.h:
```diff
@@ -0,0 +1,147 @@
+     * \param[in] vertices Vector of the simplex vertices indices (a simplex of dimension \f$q\f$ has \f$q+1\f$ vertices and the vector must be **sorted**).
+     * \param[in] sort_data If `true` vectors of vertex indices are sorted, if `false` indices are assumed to be sorted (faster).
+     */
+    Simplex(const Simplex_data_type& vertices, bool sort_data = false) : _vertices(vertices) {
```
What is `Simplex_data_type` ?

afabri committed:
indexes -> indices

afabri committed:
indexes -> indices

afabri commented on HDVF/doc/HDVF/Concepts/SparseChain.h:
```diff
@@ -0,0 +1,307 @@
+    /*!
+     * \brief Type of coefficients stored in the matrix (a model of `IntegralDomainWithoutDivision`).
+     */
+    typedef IntegralDomainWithoutDivision CoefficientRing;
```
```suggestion
    typedef unspecified_type Coefficient_ring;
```

afabri replied:
And the same for the matrix class.

afabri committed:
friend -> relates

afabri committed:
typos

afabri committed:
Remove \defgroup ChainChainComparison Compares two chains.

afabri committed:
fix \relates

afabri committed:
\relates must be outside of the class

afabri committed:
\relates must be outside of the class

afabri commented:
/force-build:v1

afabri committed:
\relates must be outside of the class

afabri commented:
/force-build:v1

umenohana13 committed:
Fixes

- Removed concept 'Ring' (replaced by the existing concept `IntegralDomainWithoutDivision`
- Fixed the proper use of words "hole" and "interval" in Hdvf_persistence iterator

umenohana13 committed:
Merge branch 'Homology-new_package-AB' of github.com:umenohana13/mine into Homology-new_package-AB

umenohana13 committed:
Update AbstractChainComplex.h

umenohana13 committed:
Fix

afabri committed:
Add Algebraic_structure_traits<Z2>

afabri committed:
vertices coordinates -> vertex coordinates

afabri committed:
Add missing '}'

afabri committed:
Add Real_embeddable_traits

afabri committed:
Add AST::Is_square

afabri committed:
It may be Uncertain<Comparison_result>

umenohana13 committed:
Update HDVF/doc/HDVF/Concepts/SparseChain.h

Co-authored-by: Andreas Fabri <andreas.fabri@geometryfactory.com>

umenohana13 committed:
Fixes

-> remaining Camelcase in Geometric_chain_complex_tools
-> small regression in Sparse_chain (incorrect end of class after reformatting)
-> small regression in Sparse_matrix after correction of indexes -> indices

umenohana13 committed:
Merge branch 'Homology-new_package-AB' of github.com:umenohana13/mine into Homology-new_package-AB

afabri commented:
/force-build:v1

afabri commented on HDVF/include/CGAL/OSM/Sparse_matrix.h:
```diff
@@ -0,0 +1,2072 @@
+     * \return The modified matrix representing the result.
+     */
+    template <typename _CT, int _CTF>
+    friend Sparse_matrix<_CT, _CTF>& del_column(Sparse_matrix<_CT, _CTF>& matrix, size_t index);
```
Can you please replace `del_`  with `remove_`

umenohana13 committed:
Fixes

-> Sparse_chain iterator warning
-> Minor fix in Z2

afabri committed:
Add two specializations of Algebraic_structure_traits for prime and not prime

afabri committed:
typos

umenohana13 committed:
Fixes

-> fixes in Hdvf_core and Hdvf_persistence code (removed abs calls -> test coefficient agains 1 / -1)
-> fixed parameters for Algebraic_structure_traits in Zp and Z2 (now compiling)
TODO: test functors

umenohana13 committed:
Update Zp.h

afabri commented on HDVF/include/CGAL/HDVF/Filtration_lower_star.h:
```diff
@@ -0,0 +1,236 @@
+     * \param[in] K Constant reference to the underlying complex.
+     * \param[in] deg_fun Function mapping vertices of `K` to their degree.
+     */
+    Filtration_lower_star(const ChainComplex& K, std::function<Degree(size_t)>& deg_fun) : Filtration_parent(K)
```
The parameter `deg_fun`  must be renamed as otherwise there is a doxygen link. See [here](https://cgal.github.io/9004/v1/HDVF/classCGAL_1_1Homological__discrete__vector__field_1_1Filtration__lower__star.html#abd5d7eac3c5509887aa5bcbf32ab2abd)

afabri committed:
typos

umenohana13 committed:
Update HDVF/include/CGAL/HDVF/Hdvf_core.h

Co-authored-by: Andreas Fabri <andreas.fabri@geometryfactory.com>

umenohana13 committed:
Fix

Changed deg_fun -> degree_function

umenohana13 committed:
Fix

Fixed all remarks mentionned in Github
(replaced all del_ with remove_)

afabri committed:
Introduce geometric traits class

afabri committed:
Added license header

afabri committed:
Add tparam

afabri committed:
Deal with 2D output

afabri committed:
Add example for Surface_mesh

afabri committed:
Add example for Surface_mesh

afabri committed:
Drop prefix example_  from example file names

afabri committed:
cleanup

afabri committed:
Start fixing test

umenohana13 committed:
Zp

-> implemented Is_invertible
-> started unit tests

afabri committed:
replace vertex_coords() with point()

afabri committed:
insert -> write

afabri committed:
insert -> write

umenohana13 committed:
Fixes

-> Fixes to integrate the new Hdvf_traits_3
-> Bug fix on Zp

umenohana13 committed:
Fixes

-> Fixes to integrate the new Hdvf_traits_3
-> Bug fix on Zp

afabri committed:
I suggest that we always return const& which corresponds to keeping cget_column()

umenohana13 committed:
Test for Zp

-> test covering Zp
-> still need to test functors

umenohana13 committed:
Fixes

-> fixed tests and examples with new Hdvf_traits_3
-> fixed tests and examples to use a default model if no argument is provided
-> fixed tests and examples to output their results to a tmp directory
-> WARNING : Sparse_matrix set to the old version (with cget_ and get_)

umenohana13 committed:
Fixes

-> Fix in Zp operator-=
-> Fixes in examples (to provide a default model if no argument is given)

afabri committed:
Add Hdvf_traits_d.h

afabri committed:
Merge cgal/main

umenohana13 committed:
Fixes

-> fixed small regressions in Icosphere_object_io
-> added a subdivision parameter for the bounding sphere in Geometric_chain_complex_tools / simplicial_chain_complex_bb
-> added a default parameter for dual_hdvf_simplicial (when no model is provided)

umenohana13 committed:
Test Hdvf_core

-> Finished a rough test for Hdvf_core (testing the Hdvf build for a simple model)
-> Added compare and operator== methods to Hdvf_core

umenohana13 committed:
Fixes

-> created a vector _points of points in Cubical_chain_complex
-> fixed point() and points() accordingly
-> started to generalize the point dimension (created a Hdvf_traits_2)

TODO: improve "to_point" functions (used to generate a Point in the right dimension)

umenohana13 commented on HDVF/include/CGAL/HDVF/Mesh_object_io.h:
```diff
@@ -0,0 +1,658 @@
+    friend Io_node_type & operator+(const Io_node_type &v1, const Io_node_type &v2)
+    {
+        assert(v1.size() == v2.size()) ;
+        Io_node_type &tmp = *new(Io_node_type)(v1.size(),0) ;
```
Fixed (by adding a _points vector in Cubical_chain_complex.cpp)
But I'm bothered: how to generate points properly according to the dimension: Point_2 and Point_3 are created by given x, y ... but Point_d are created using an iterator...
I've tried to create overloaded "to_point" functions, taking the vector of doubles of coordinates and returning Point in the right dimension. This function is called by "compute_point" in the class. I thought that overloading would solve the problem. But as soon as I try to add the Point_2 version, I get compilation errors...

afabri committed:
If to_point() is in the traits you can call Traits().to_point(vec)

umenohana13 committed:
Fix (points / cubical chain complex)

-> changed Point output to use a constexpr in Cubical_chain_complex -> point
-> removed to_point functions

umenohana13 committed:
Traits in various dimensions

-> Fixed point/points in dimension 2, 3, d
-> Created a Point3 in all the Traits with a static to_point3 method to handle 3D export "cleanly"
-> Updated classes accordingly

umenohana13 committed:
Moved read_nodes

In order to load higher dimensional simplicial data, we will load a .simp (through the Mesh_object_io) + vertices coordinates (through a .nodes file - in tetgen format). So I moved the read_nodes function to Mesh_object_io (also used from Tet_object_io)

umenohana13 committed:
Test Hdvf_traits_d <4>

-> created 4D model (Klein bottle)
-> created hdvf_simplicial_4
PROBLEM: cannot compile but errors come from function_objects_cartesian and Hyperplane...?

afabri committed:
dD needs Eigen, other fixes to make hdvf_simplicial_4.cpp compile

umenohana13 committed:
4D simplicial OK

-> tested hdvf_simplicial_4 over the Klein bottle (results are coherent).
-> vtk export OK (4D space is automatically projected to x,y,z)

I'm preparing other 4D models to test simplicial and cubical results
TODO
-> and it can be nice to integrate a mapping function (R^4->R^3) into the to_point3 method so that we can change the view into R^4

umenohana13 committed:
Fix

Fix on the torus4D model

afabri committed:
Add <Traits>  so that examples compile

afabri committed:
Some fixes for VC++

umenohana13 committed:
dD export improvement

-> changed to_point3 functions in Hdvf_traits_ to functional style (to let the use chose / define the right projection)
-> provide several "default" methods for dD (from 3 orthonormal vectors / from the pca frame)

umenohana13 committed:
HDVF duality

-> simplification of Geometric_chain_complex_tools (simplicial_chain_complex_bb now receives a mesh_object_io to provide a support for "heterogeneous" simplicial complexes - with edges and non manifold faces)
This function relies on tetgen (which will be external as soon as I've finish writing a .poly exporter)
TODO
-> the function will be overloaded to receive a Mesh_surface. In this case, all computations will be carried out into CGAL with Delaunay_constrainted_triangulations_3

afabri committed:
Use Epick_d functors

afabri committed:
doc bug

afabri committed:
doc bug

afabri committed:
At some place it needs the concept and not the class name.  But needs discussioon for some

afabri committed:
Move doc of Cell, so that Filtration_core doc shows up;  boolean -> Boolean

afabri commented:
/force-build:v1

afabri committed:
replace concept with local typedef

afabri committed:
Add a typedef

afabri commented on HDVF/include/CGAL/HDVF/Hdvf.h:
```diff
@@ -0,0 +1,1474 @@
+ */
+
+template<typename ChainComplex>
+class Hdvf : public Hdvf_core<ChainComplex, OSM::Sparse_chain, OSM::Sparse_matrix> {
```
How can that compile without tparams for `OSM::Sparse_chain`

afabri committed:
cleanup

afabri committed:
cleanup

afabri committed:
cleanup

afabri committed:
In the documentation of the concept we cannot put a model

afabri committed:
In the documentation of the concept we cannot put a model

umenohana13 committed:
Traits_d and Duality devs

-> Added (and fixed) dim_d -> dim_3 projectors in Traits_d (among them, PCA projector)
-> Started to add CGAL constrained Delaunay triangulation for duality

afabri committed:
Merge remote-tracking branch 'cgal/main' into Homology-new_package-AB

afabri commented:
/force-build:v1

umenohana13 committed:
Triangulation_3_io

-> added a Triangulation_3_io class (to load Triangulation_3 tetraedrizations)
-> simple test passed

umenohana13 committed:
Merge branch 'Homology-new_package-AB' of github.com:umenohana13/mine into Homology-new_package-AB

umenohana13 committed:
CGAL CDT3 dualization

-> provide a fully CGAL duality approach (using CDT3)
-> PROBLEM: CDT3 fails (on a simple off mesh: two rings inside a surrounding icosphere. For tests, the mesh is exported to "tmp/test_bounded.off".

umenohana13 committed:
Duality with CDT3 CGAL

-> Fixed issues with the test of CDT3
-> PROBLEM remains (CDT3 fails)

afabri committed:
Use Exact_predicates_inexact_constructions_kernel

afabri commented on HDVF/include/CGAL/HDVF/Icosphere_object_io.h:
```diff
@@ -0,0 +1,133 @@
+*/
+
+template <typename Traits>
+class Icosphere_object_io : public Mesh_object_io<Traits>
```
Note that CGAL has the function [`make_icosahedron()`](https://cgal.github.io/9115/v1/BGL/group__PkgBGLGeneratorFct.html#ga03d4d6b1b7cb4ef6c460487240e6bb10)

umenohana13 replied:
Yes, in the CGAL version, I use this function to generate the icosahedrons Surface_mesh. But here, I need to create a Mesh_object_io ... and anyway as soon as possible, this tetgen version will disappear:)

umenohana13 committed:
Duality with CGAL CDT3

-> Finished the first version (run dual_hdvf_simplicial_cgal with local working directory).
-> OK for a simple mesh (three_triangles)
-> but in spite of faces indices provided (that should be preserved), CDT3 remeshes a face of the initial data and thus duality fails (as the CDT3 needs to CONTAIN the initial mesh).

umenohana13 committed:
Merge branch 'Homology-new_package-AB' of github.com:umenohana13/mine into Homology-new_package-AB

umenohana13 commented:
The word "**cell**" is used in Abstract_simplicial_chain_complex and Simplicial_chain_complex to speak about simplices: this is actually the only way to have a coherent description of complexes whatever their type (simplicial, cubical and later on, linear - such as combinatorial maps).

umenohana13 committed:
Fixes

-> Simplex_data_type now public and commented
-> Added Cubical_chain_complex in AbstractChainComplex models

umenohana13 committed:
Fix

Sorry, Cubical_chain_complex is a model of GeometricChainComplex...

umenohana13 committed:
Fixes

-> changed all parent classes names to Base
-> removed useless Base:: and replaced them by using ...

umenohana13 commented on HDVF/include/CGAL/HDVF/Hdvf.h:
```diff
@@ -0,0 +1,1477 @@
+     * \param[in] chain The cycle to annotate in the homology basis.
+     * \param[in] dim Dimension of the cycle.
+     */
+    typename Base::Column_chain get_annotation(typename Base::Column_chain chain, int dim) const
```
Done
Commit: 22931c62ba6830eecd585afe2a812f84dfa2fa63

umenohana13 committed:
Fix

Changed simplicial_chain_complex_bb to dualize_complex

umenohana13 committed:
Fixes

-> fixed the confusion between class and concept in the concept of SparseMatrix and SparseChain
-> removed useless COLUMN/ROW and add \pre
-> renamed variables in corresponding classes accordingly

umenohana13 committed:
Fixes

PR fixes (following remarks in "encore un commentaire" email)

umenohana13 committed:
Fixes (Z2 and Zp)

-> fixed all remarks from the email "Zp et Z2"

umenohana13 committed:
Fixes

Fixes on duality (Complex_duality_data is now the return type of all dualize_complex methods)

TODO: returns references (to save memory and copying type) which may lead to memory leaks...

afabri committed:
Make that degree_function has no link

afabri committed:
fix link

umenohana13 committed:
Duality fix

-> created a class `Complex_duality_data_t` containing two pointers on the complex/sub_chain_complex_mask built by `dualize_complex`
-> `dualize_complex` now take a reference on such an object to save its output. So the object must be built (deleted) outside the method and deleting it also frees the corresponding chain_complex and sub_chain_complex_mask

umenohana13 committed:
Merge branch 'Homology-new_package-AB' of github.com:umenohana13/mine into Homology-new_package-AB

umenohana13 committed:
Merge branch 'Homology-new_package-AB' of github.com:umenohana13/mine into Homology-new_package-AB

Corrected `Complex_duality_data_t` class (documentation and delete over the copy constructor).

umenohana13 committed:
CGAL CDT3: help needed

-> started integrating faces subdivision in CDT3 with plc_face_id, but visit of faces constraints fails

afabri committed:
Add @param

afabri commented:
/force-build:v1

albert-github commented:
@afabri 

The link you provided leads me to:

<img width="557" height="60" alt="image" src="https://github.com/user-attachments/assets/aac97a9c-0bae-432f-b678-310a06bd4c11" />

in the "Protected Member Functions" section.
I don't see a link here or do you mean why is this function here at all?

afabri committed:
Fix dualize()

umenohana13 committed:
Duality fixes

-> changed Complex_duality_data_t: it now contains two shared_ptr
-> dualize_complex functions all return a Complex_duality_data_t
-> all examples corrected
-> added dual_hdvf_cubical example (for cubical data)

afabri committed:
Add @param

umenohana13 committed:
write_VTK fixes

-> created and documented write_VTK functions for `GeometricChainComplex` models and for chains in such complexes.
-> removed documentation for chain_complex_to_vtk and chain_to_vtk in `GeometricChainComplex` models
-> tested in test_hdvf_examples

umenohana13 committed:
Merge branch 'Homology-new_package-AB' of github.com:umenohana13/mine into Homology-new_package-AB

afabri committed:
Move Z2 and Zp to Number_types

afabri committed:
Move Z2 and Zp to Number_types

afabri committed:
move files in another doxygen group

afabri committed:
calculate -> compute

afabri committed:
Add links fdor Zp and Z2

afabri committed:
Add a doxygen group for IO classes

afabri committed:
Add package description

afabri committed:
typos

afabri committed:
typos

afabri committed:
typos

afabri committed:
typos

afabri committed:
typos

umenohana13 committed:
Typos

umenohana13 committed:
Implemented dimension restriction

HDVFs can now be computed along a single dimension q (actually in this case, pairs must be added in dimensions q/q+1 and q-1/q).

All functions of HDVF_core and Geometric_chain_complex have been updated.

TODO: implement dimension restriction for persistence and duality (feasible).

Speedup must be estimated, but for cubical models (including a large number of cells of dimension 3), the speedup is important.

umenohana13 committed:
Dimension restriction for persistence

Implemented dimension restriction for persistence (together with a test for cubical complexes - largest complexes used for tests)

umenohana13 committed:
Exceptions in Abstract_simplicial_chain_complex

Exceptions OK + documentation

umenohana13 committed:
Exceptions Cub_object_io

Revised all exceptions in Cub_object_io (+documentation)

umenohana13 committed:
Fixes

-> added a comment in the documentation of Surface_mesh_io and Mesh_object_io on the unicity of vertices (no duplicates)
-> fixed trailing whitespaces ?

umenohana13 committed:
Fixes

Trailing whitespaces

umenohana13 committed:
Fixes

Trailing whitespaces

umenohana13 committed:
Fixes

Trailing whitespaces

umenohana13 committed:
Fixes

Trailing whitespaces

afabri commented:
/force-build:v1

sloriot committed:
fix doc warnings

sloriot committed:
add missing throw

afabri commented:
/force-build:v1

sloriot committed:
fix backticks

sloriot committed:
typos

sloriot committed:
clean up

umenohana13 committed:
Fixes

umenohana13 committed:
Merge branch 'Homology-new_package-AB' of github.com:umenohana13/mine into Homology-new_package-AB

afabri committed:
Change one exception documentation

umenohana13 committed:
Fixes

Fixes for gcc / Linux compiler

umenohana13 committed:
Merge branch 'Homology-new_package-AB' of github.com:umenohana13/mine into Homology-new_package-AB

umenohana13 committed:
Fixes

Fixes on dimension restriction for persistence (infinite holes extraction)

umenohana13 committed:
Added "is_valid_pair" operations

For all HDVF operations (A, R, M, W, MW) added a correponding operation to check the validity of a pair of argument cells provided by the user.

umenohana13 committed:
z and co_z functions

Added z and co_z functions to the API (functions defining fundamental cycles/co-cycle)
Necessary to related HDVFs to tri-partitions

umenohana13 committed:
Methods hd, htdt

Added protected methods computing hd(sigma), <hd(sigtma), tau> and same for h^*d^*.
These matrices are not store (for efficiency) so we must provide efficient computations for "advanced users" developping topological algorithms -> therefore the methods are protected

umenohana13 committed:
Fix

umenohana13 committed:
Fixes

umenohana13 committed:
Fixes

-> tried to fix problem on set_column and set_row (git compilation)

sloriot committed:
void cyclic dependency

+missing typenames

umenohana13 committed:
Fixes

-> fixed a warning in Cub_object_io (function without return)
-> fixed errors on f_x, f_y, f_z

sloriot committed:
add dependencies

umenohana13 committed:
Fix

Removed static in Z2 / Zp

umenohana13 committed:
Merge branch 'Homology-new_package-AB' of github.com:umenohana13/mine into Homology-new_package-AB

umenohana13 committed:
Fixes

-> removed useless f_x, f_y, f_z functions in Filtration_lower_star
-> simplified examples in persistent_hdvf_simplicial (various versions are commented)

umenohana13 committed:
Fixes

umenohana13 committed:
Update Cubical_chain_complex.h

umenohana13 committed:
Update Cubical_chain_complex.h

umenohana13 committed:
Exception handling, cleaning

-> Fixed all exception handling in Abstract_simplicial_chain_complex, Simplicial_chain_complex, Cubical_chain_complex
-> Added CGAL::IO::write_VTK function for VTK exports (chain complexes and chains)

umenohana13 committed:
Fixes

-> updated Cub_object_io and Mesh_object_io
-> moved member data to private
-> created access methods
-> documentation OK (including exceptions)

umenohana13 committed:
Surface_mesh_io

Removed documentation of private member variables

umenohana13 committed:
Fixes

umenohana13 committed:
Fixes

-> Documented Tet_object_io
-> Moved to protected all internal functions

umenohana13 committed:
Fixes

umenohana13 committed:
Fixes

umenohana13 committed:
Fixes

umenohana13 committed:
Exception and rework of Hdvf

-> started integrating the is_invertible test in Hdvf operations
-> checked is_invertible and inverse in Z2 and Zp
-> test operation R in test_hdvf

umenohana13 committed:
Hdvf / Hdvf_core rework

-> replaced partial tests by .is_invertible / .inverse in R, M and W operations
-> test_hdvf covers these operations
-> finished is_invertible and inverse in Z2 and Zp

umenohana13 committed:
Hdvf_core operation A

-> corrected operation A and all find_pair methods (to use .is_invertible and .inverse)
-> tested in test_hdvf

umenohana13 committed:
Fixes

umenohana13 committed:
OperationMW

-> operation corrected with is_invertible/inverse
-> tested (+all find_pair methods, is_valid_pair_for_MW) in test_hdvf

umenohana13 committed:
Fixes

Added CGAL_precondition to operations

umenohana13 committed:
Fixes

umenohana13 committed:
Fixes

umenohana13 committed:
Update Hdvf_persistence.h

-> Reviewed exceptions
-> Used is_invertible to test possible A operations

afabri committed:
typos

umenohana13 committed:
Hdvf_persistence + fixes

-> reviewed Hdvf_persistence
-> started test_hdvf_persistence
-> fixed remarks by Andreas in Concepts

afabri committed:
Do not explain what an assignemnt means

afabri committed:
Do not explain what an assignemnt means

umenohana13 committed:
Fixes (Hdvf api) + tests

-> corrected hd and dh technical methods (provided as protected) - tested in test_hdvf
-> added two psc_flags methods returning PSC_flags in 1 / any dimension
-> started test_hdvf_persistence

umenohana13 committed:
Merge branch 'Homology-new_package-AB' of github.com:umenohana13/mine into Homology-new_package-AB

umenohana13 committed:
Fix

umenohana13 committed:
Hdvf_persistence

-> added dedicated read/write_hdvf_reduction
-> add dedicated compare
-> test_hdvf_persistence
-> fixes

umenohana13 committed:
Fixes

umenohana13 committed:
Fixes

umenohana13 committed:
Trailing whitespaces fixes

umenohana13 committed:
Hdvf_persistence exceptions

-> Exceptions handled

umenohana13 committed:
.hdvf file format prefix

-> added a prefix to the file format distinguish between Hdvf_core / Hdvf_persistence / Hdvf_duality saves
-> updated test data, Hdvf_core and Hdvf_persistence accordingly
-> tests OK

umenohana13 committed:
API update

-> tested the copy constructor
-> added constructor for Hdvf and Hdvf_core from PSC_flags (TODO: build the associated reduction)
-> updated the write function (created a write_flags method)

umenohana13 committed:
Fix

umenohana13 committed:
Fixes

umenohana13 committed:
Fixes Hdvf

-> overloaded all operations (A, R, M, W, MW) to accept Cell_pair
-> changed the hdvf_simplicial to use Surface_mesh_io loader for meshes instead of Mesh_object_io
-> created hdvf_abstrat_simplicial example for Abstract_simplicial_chain_complex and inserted examples of M, W, MW operations

umenohana13 committed:
MW fix + beginning of LU for Sparse_matrix

umenohana13 committed:
Full_LU for a ring

-> finished coding full LU decomposition over a ring for OSM::Sparse_matrix
-> tested

umenohana13 committed:
Full_lu + hdvf from flags

-> Finished and tested Full_lu for Sparse_matrix (with corresponding solve, inverse ... methods)
-> Added a construction of Hdvf_core and Hdvf from flags
--> Added extraction/restriction/fill sub matrix methods
--> Added a reconstruction of the reduction from the flags
--> Tested

umenohana13 committed:
Fixes

umenohana13 committed:
Fixes

umenohana13 committed:
Fix on MW

-> Fixed a bug on MW computation
-> Operations have now been tested extensively  (current internship) and are fully operational.

umenohana13 committed:
Class Full_lu

-> moved class Full_lu to Full_lu.h
-> updated code accordingly
-> finished determinant and is_invertible
-> tested

umenohana13 committed:
Fixes

-> corrected missing const in write_VTK

umenohana13 committed:
Fixes

-> fixed undocumented variables in the documentation of Hdvf and Hdvf_core new methods
-> Fixed Hdvf_core copy constructor (for reduction dimension variables)
-> changed get_id() method to id() and added it to the AbstractChainComplex concept

sloriot committed:
Merge remote-tracking branch 'cgal/main' into umenohana13:Homology-new_package-AB

sloriot committed:
fix dependencies

umenohana13 committed:
Started revision of Hdvf_duality

-> Added dedicated versions of read/write and compare methods
-> Tested
-> Added set methods to Sub_chain_complex_mask

umenohana13 committed:
Hdvf_duality review

-> added a verification of empty complexes (also added in all Hdvf classes)

umenohana13 committed:
Revision Hdvf_duality

-> properly test if HDVFs over K and L-K are perfect before computing an Alexander duality pairing (this is a pre-requisite)
-> various fixes
-> added proper is_null methods for sub_sparse_matrix

umenohana13 committed:
Update Traits

-> added comments for projection methods
-> revised comments

umenohana13 committed:
Intermediate version for Sparse_matrix

sloriot committed:
fix warnings and compilation error

umenohana13 commented:
Questions for May 20th:
- Names / tparams related to Sparse_matrix (Sparse_matrix_core and Sparse_matrix curryfication providing Sparse_matrix_type and corresponding underlying Sparse_chain_type)
- How to describe this in the concepts?
- is_invertible() and inverse() for integers?
- Hdvf_persistence and Hdvf_duality must take Complexes built over Sub_sparse_matrix... is there a way to guarantee / enforce it?

afabri commented:
/force-build:v0

umenohana13 committed:
Sparse_matrix taking Sparse_chain as tparam

-> Created Sparse_matrix_core taking Sparse_chain as tparam
-> created Sparse_matrix structure, curryfication of Sparse_matrix_core, taking the Sparse_chain type and providing 2 templates (for corresponding sparse chain and sparse matrix)
-> updated most classes + documentation
-> started updating tests and examples

afabri commented:
/force-build:v0

umenohana13 committed:
Fixes

-> fixed doc
-> fixed changes wrt Sparse_matrix

afabri commented:
/force-build:v0

afabri committed:
small cleanup

afabri committed:
Introduce Is_invertible and Inverse

afabri committed:
-1 is also invertible

afabri committed:
Add AS<Zp>::Invert

afabri committed:
Invert -> Inverse

afabri committed:
Call free function is_invertible() and inverse()

afabri committed:
trailing whitespace

umenohana13 committed:
Fixes (integrating new sparse matrices)

umenohana13 committed:
Tried using is_invertible and inverse

-> Full_lu
-> Hdvf
-> Couldn't understand how to call generic functions (for Zp and int)?

afabri committed:
Prefix some calls with CGAL:: so that the compiler does not mix up with the local member function

afabri committed:
line is too long

afabri committed:
Add #include

umenohana13 committed:
New Sparse_matrix structure

-> Sparse matrices now receive the sparse chain type as a tparam
-> Updated all classes accordingly
-> All tests are OK
-> Concepts updated (need checks especially for AbstractSimplicialChainComplex - how to describe the Sparse_matrix structure passed as argument...)

umenohana13 committed:
Update Sparse_matrix depending on Sparse_chain

-> review all the code / examples / tests to comply with new structure
@TODO: check the documentation (related modifications are commented with warnings)

afabri commented:
/force-build:v0

afabri commented on HDVF/doc/HDVF/dependencies:
```diff
@@ -0,0 +1,9 @@
+Number_types
+Circulator
+Stream_support
+Surface_mesh
```
```suggestion
Surface_mesh
Triangulation_3

```

afabri replied:
We then get a link [here](https://cgal.github.io/9004/v0/HDVF/classCGAL_1_1Homological__discrete__vector__field_1_1Triangulation__3__io.html)

afabri commented on HDVF/include/CGAL/OSM/Sparse_matrix.h:
```diff
@@ -0,0 +1,2232 @@
+     *
+     * \brief Comparison of two COLUMN matrices.
+     */
+    template <typename _CT, template <typename, int> typename SCT>
```
You have to write what `SCT`   means

afabri commented on HDVF/include/CGAL/OSM/Sparse_matrix.h:
```diff
@@ -0,0 +1,2232 @@
+
+    /** \relates Sparse_matrix_core
+     *
+     * \brief Comparison of a COLUMN  and a ROW matrix.
```
```suggestion
     * \brief Comparison of a `COLUMN`  and a `ROW` matrix.
```

umenohana13 committed:
Fix
